lib/ai/embeddings/embeddings_test.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2023 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package embeddings_test
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/internalutils/stream"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/ai"
+	aiembeddings "github.com/gravitational/teleport/lib/ai/embeddings"
+	"github.com/gravitational/teleport/lib/backend/memory"
+	"github.com/gravitational/teleport/lib/services/local"
+)
+
+// MockEmbedder returns embeddings based on the sha256 hash function. Those
+// embeddings have no semantic meaning but ensure different embedded content
+// provides different embeddings.
+type MockEmbedder struct {
+}
+
+func (m MockEmbedder) ComputeEmbeddings(_ context.Context, input []string) ([]ai.Vector64, error) {
+	result := make([]ai.Vector64, len(input))
+	for i, text := range input {
+		hash := sha256.Sum256([]byte(text))
+		vector := make(ai.Vector64, len(hash))
+		for j, x := range hash {
+			vector[j] = 1 / float64(int(x)+1)
+		}
+		result[i] = vector
+	}
+	return result, nil
+}
+
+func TestNodeEmbeddingGeneration(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	clock := clockwork.NewFakeClock()
+
+	// Test setup: crate a backend, presence service, the node watcher and
+	// the embeddings service
+	bk, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clock,
+	})
+	require.NoError(t, err)
+
+	embedder := MockEmbedder{}
+	presence := local.NewPresenceService(bk)
+	embeddings := local.NewEmbeddingsService(bk)
+
+	processor := aiembeddings.NewEmbeddingProcessor(&aiembeddings.EmbeddingProcessorConfig{
+		AiClient:     &embedder,
+		EmbeddingSrv: embeddings,
+		NodeSrv:      presence,
+		Log:          logrus.WithField(trace.Component, "test"),
+		Jitter:       retryutils.NewSeventhJitter(),
+	})
+
+	done := make(chan struct{})
+	go func() {
+		err := processor.Run(ctx, 100*time.Millisecond)
+		require.ErrorContains(t, err, "context canceled")
```

Comment:
Can't use require in a goroutine like this.

Reply:
I keep forgetting about this 🤦 I'll replace it with `assert` that work in this case.